src/cljs/rems/administration/create_form.cljs

```diff
@@ -193,15 +206,42 @@
 (defn- validate-options [options languages]
   {:field/options (apply merge (mapv #(validate-option %1 %2 languages) options (range)))})
 
-(defn- validate-field [field index languages]
+(defn- field-option-keys [field]
+  (set (map :key (:field/options field))))
+
+(defn- validate-only-if-field [field visibility fields]
+  (if (-> visibility :visibility/field)
+    (if-let [referred-field (find-first (comp #{(get-in visibility [:visibility/field :field/id])} :field/id) fields)]
+      (if-not  (supports-options? referred-field)
+        {:field/visibility {:visibility/field :t.form.validation/invalid-value}}
+        (if-not (-> visibility :visibility/field :field/id)
+          {:field/visibility {:visibility/field :t.form.validation/invalid-value}}
+          (if-not (empty? (:visibility/values visibility))
+            (when-not (some (field-option-keys referred-field) (:visibility/values visibility))
+              {:field/visibility {:visibility/values :t.form.validation/invalid-value}})
+            {:field/visibility {:visibility/values :t.form.validation/required}})))
+      {:field/visibility {:visibility/field :t.form.validation/required}})
+    {:field/visibility {:visibility/field :t.form.validation/required}}))
+
+(defn- validate-visibility [field fields]
+  (when-let [visibility (:field/visibility field)]
+    (case (:visibility/type visibility)
+      :always nil
+      :only-if (validate-only-if-field field visibility fields)
+      nil {:field/visibility {:visibility/type :t.form.validation/required}}
+      {:field/visibility {:visibility/type :t.form.validation/invalid-value}})))
+
+(defn- validate-field [field index languages fields]
```

Comment:
if this was `validate-fields`, wouldn't need to pass `field`, `index` and `fields` separately

Reply:
Sounds like a good refactoring to do.

Reply:
Done.